src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Copyright 2019, Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.querying.evaluators;
+
+import com.yahoo.bullet.query.expressions.Expression;
+import com.yahoo.bullet.query.expressions.Operation;
+import com.yahoo.bullet.record.BulletRecord;
+import com.yahoo.bullet.typesystem.Type;
+import com.yahoo.bullet.typesystem.TypedObject;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Evaluators do the work of expressions.
+ *
+ * Each expression is built into a corresponding evaluator. Note, evaluators are constructed after a query has been
+ * initialized so assume all expressions are valid.
+ *
+ * Evaluators are evaluated given a BulletRecord and will throw exceptions on any errors. These errors are virtually all
+ * from some form of type check.
+ *
+ * Also, note the type cast in evaluator. For primitives, this acts as how you think it would, but for lists and maps, it
+ * will cast their elements/values.
+ */
+public abstract class Evaluator {
+    @FunctionalInterface
+    public interface BinaryOperator {
+        TypedObject apply(Evaluator left, Evaluator right, BulletRecord record);
+    }
+
+    @FunctionalInterface
+    public interface UnaryOperator {
+        TypedObject apply(Evaluator evaluator, BulletRecord record);
+    }
+
+    @FunctionalInterface
+    public interface NAryOperator {
+        TypedObject apply(List<Evaluator> evaluator, BulletRecord record);
+    }
+
+    // For testing only
+    @Getter(AccessLevel.PACKAGE)
+    protected Type type;
+
+    static final Map<Operation, BinaryOperator> BINARY_OPERATORS = new EnumMap<>(Operation.class);
+    static final Map<Operation, BinaryOperator> BINARY_ANY_OPERATORS = new EnumMap<>(Operation.class);
```

Comment:
What's the point of distinguishing ANY and ALL? Even in BinaryExpression, there's some enum for ANY and ALL and you pick from these corresponding operators but what's the point? Might as well just pick from BINARY_OPERATORS? If this is a BQL thing, we should probably not push it all the way down here.